### PR TITLE
Full implementation

### PR DESCRIPTION
**Pull request by Burgyn**
# Ocelot ETag Caching

## Basic

The Ocelot ETag Caching library adds support for ETag caching to the Ocelot API Gateway. ETag caching is a mechanism that allows a client to cache data and the next time the same data is requested, the client can verify that the data is still up to date. If the data is still current, the server will return a status of `304 Not Modified` and the client will use the cached data. If the data is not up-to-date, the server returns the data and the client caches it.

The idea is that the server adds two important headers to the response:

- `ETag` - data identifier (randomly generated value)
- `cache-control` - identifier that the data can be cached. Contains only the value `private` (⚠️ beware it must not be `public`, because then the data can remain cached anywhere.) It also does not contain `max-age` because in this case the client would not verify the data on the server (for a given amount of time. Occasionally this may be OK).

If the client (browser) finds these two headers in the response, it adds the `If-None-Match` header with the `ETag` value in the next request to the server with the same path and the server knows if the data is still up to date based on this value. If they are up to date it returns a status of `304 Not Modified` and the client uses the data from the cache.
**👌 The server does not send the data. This saves resources and bandwidth**

On the client, this works automatically because this behavior is defined in the HTTP specification (no need to do anything).

## Implementation

The implementation is based on Ocelot middleware. All caching will be done in this middleware and nothing will be needed on the service side. The data itself is not cached, but only its identifier (ETag), based on which the data is verified to be up-to-date.

We use `IOutputCacheStore` to store ETags and invalidate them.

### Ocelot configuration

```json
{
    "Routes": [
        {
            "Key": "getAllProducts",
            "DownstreamPathTemplate": "/api/producsts/",
            "UpstreamPathTemplate": "/products/",
            "CachePolicy": "getAllProduct",
            ...
        },
        {
            "Key": "getProduct",
            "DownstreamPathTemplate": "/api/producsts/{id}",
            "UpstreamPathTemplate": "/products/{id}",
            "CachePolicy": "getProduct",
            ...
        },
        {
            "Key": "deleteProduct",
            "DownstreamPathTemplate": "/api/producsts/{id}",
            "UpstreamPathTemplate": "/products/{id}",
            "InvalidateCachePolicies": ["getProduct", "getAllProducts"],
            ...
        }
    ]
}
```

### `Program.cs` configuration

```csharp
builder.Services.AddOcelotETagCaching((c) =>
    {
        // 👇 Add ETag caching policies
        // 👇 Simple policy with Expire and tag templates
        c.AddPolicy("getAllProducts", p =>
        {
            p.Expire(TimeSpan.FromMinutes(5));
            p.TagTemplates("products:{tenantId}", "all", "tenantAll:{tenantId}");
        });

        // 👇 Policy with custom cache key, etag generator and custom cache control
        c.AddPolicy("getProduct", p =>
        {
            p.Expire(TimeSpan.FromMinutes(5));
            p.TagTemplates("product:{tenantId}:{id}", "tenant:{tenantId}:all", "all");

            p.CacheKey(context => context.Request.Headers.GetValues("id").FirstOrDefault()); // 👈 Custom cache key
            p.ETag(context => new($"\"{Guid.NewGuid()}\"")); // 👈 Custom etag
            p.CacheControl(new() { Public = false }); // 👈 Custom cache control
            p.StatusCode(222); // 👈 Custom status code
            p.AddPolicy<MyCustompolicy>(); // 👈 Custom policy
        });
    }
);

...

app.UseOcelot(c =>
{
    // 👇 Add etag caching middleware
    c.AddETagCaching();
}).Wait();

app.Run();
```

## Redis

By default is used `InMemoryCacheStore` but you can use `Redis` as well.

```csharp
builder.Services.AddStackExchangeRedisOutputCache(options =>
{
    options.Configuration = 
        builder.Configuration.GetConnectionString("MyRedisConStr");
    options.InstanceName = "SampleInstance";
});
```

### CODE REVIEW

**Review by majco333 — APPROVED** [review]

**Review by satano — CHANGES_REQUESTED** [review]

File: src/Kros.Ocelot.ETagCaching/ETagCachePolicyBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+namespace Kros.Ocelot.ETagCaching;
+
+/// <summary>
+/// Provides helper methods to create custom policies.
+/// </summary>
+public sealed class ETagCachePolicyBuilder
+{
+    private readonly List<IETagCachePolicy> _policies = [];
+
+    internal ETagCachePolicyBuilder(bool excludeDefaultPolicy = false)
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/ETagCachePolicyBuilder.cs:14]: To ako si to predstavujem je presne tak ako to popisuješ. 99% (vymyslené číslo 😉) prípadov je o tom, že používam default policy, ktorej budem pridávať `TagTemplates`. Tá možnosť bez "default" je tam hlavne pre prípad, že v niektorom prípade budeme potrebovať niečo úplne iné.

File: src/Kros.Ocelot.ETagCaching/ETagCachingMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        return next();
+    }
+
+    private bool TryGetPolicy(HttpContext context, [NotNullWhen(true)] out IETagCachePolicy? policy)
+    {
+        var downstreamRoute = context.Items.DownstreamRoute();
+
+        if (!string.IsNullOrWhiteSpace(downstreamRoute.Key) && _routeOptions.TryGetValue(downstreamRoute.Key, out var route))
+        {
+            policy = _cachingOptions.Value.GetPolicy(route.CachePolicy);
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/ETagCachingMiddleware.cs:36]: Ak nebudeš silno tlačiť na tvoje riešenie, tak by som rád ostal pri tomto riešení.
Je to síce možno trochu overkill, ale za mňa to výrazne zjednoduší implementáciu. Nemusím na niečo myslieť, nemusím robiť rozhodovanie či niečo áno alebo nie. ...

Viem, že na úrovni famerworkov sa takéto optimalizácie zvyknú robiť, aby to bolo pre koncového vývojára čo najefektívnejšie. Ale ja by som tu uprednostnil tú priamočiarosť. Teda aspoň podľa mňa je to priamočiarejšie.

File: src/Kros.Ocelot.ETagCaching/ETagCachingOptions.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    /// <summary>
+    /// Adds a policy to the ETag caching.
+    /// </summary>
+    /// <param name="policyName">This name shoud be use for downstream path configuration.</param>
+    /// <param name="builder">Policy builder.</param>
+    /// <param name="excludeDefaultPolicy">If true, default policy will not be added.</param>
+    /// <exception cref="ArgumentException">If policy with the same name already exists.</exception>
+    public ETagCachingOptions AddPolicy(
+        string policyName,
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/ETagCachingOptions.cs:24]: Nemá. Opravujem.

File: README.md [hunk 4]
```diff
@@ -75,9 +71,11 @@ builder.Services.AddOcelotETagCaching((c) =>
             p.Expire(TimeSpan.FromMinutes(5));
             p.TagTemplates("product:{tenantId}:{id}", "tenant:{tenantId}:all", "all");
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply README.md:72]: Pridal som tam taký ľahký popis.

File: src/Kros.Ocelot.ETagCaching/Policies/TagTemplatesPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+internal sealed class TagTemplatesPolicy(string[] tagTemplates) : IETagCachePolicy
+{
+    private readonly string[] _tagTemplates = tagTemplates;
+
+    public ValueTask CacheETagAsync(ETagCacheContext context, CancellationToken cancellationToken)
+    {
+        foreach (var tagTemplate in _tagTemplates)
+        {
+            var tag = new StringBuilder(tagTemplate);
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/Policies/TagTemplatesPolicy.cs]: 👍

File: src/Kros.Ocelot.ETagCaching/Policies/CacheEntryExtraPropsPolicy.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿namespace Kros.Ocelot.ETagCaching.Policies;
+
+internal sealed class CacheEntryExtraPropsPolicy(IEnumerable<KeyValuePair<string, object>> extraProps)
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/Policies/CacheEntryExtraPropsPolicy.cs:3]: Neviem čo mi šlo vtedy hlavou 🤔

File: src/Kros.Ocelot.ETagCaching/Policies/EnableCachePolicy.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+﻿namespace Kros.Ocelot.ETagCaching.Policies;
+
+internal sealed class EnableCachePolicy : IETagCachePolicy
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/Policies/EnableCachePolicy.cs:3]: Máš pravdu. Zabudol som ju pridať do buildera.

Ona v princípe slúži na disablovanie policies ak by boli pridané nejak hromadne a človek ju chce pre špecificky endpoint zrušiť. Ale toto nepredpokladám, že sa bude používať.
Pridal som to do buildera a nechávam to zatiaľ tam.

File: src/Kros.Ocelot.ETagCaching/Policies/HttpHeadersHelper.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Kros.Ocelot.ETagCaching.Policies;
+
+internal static class HttpHeadersHelper
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/Policies/HttpHeadersHelper.cs]: fixed

File: src/Kros.Ocelot.ETagCaching/Policies/DefaultPolicy.cs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            context.AllowNotModified = true;
+        }
+
+        return ValueTask.CompletedTask;
+    }
+
+    public ValueTask ServeNotModifiedAsync(ETagCacheContext context, CancellationToken cancellationToken)
+    {
+        context.StatusCode = System.Net.HttpStatusCode.NotModified;
+        AddCacheHeaders(context.CachedResponseHeaders,context.ETag.ToString());
```
[the remark this answers is not in the document]
Comment (Burgyn) [reply src/Kros.Ocelot.ETagCaching/Policies/DefaultPolicy.cs]: fixes

**Review by satano — APPROVED** [review]

**Pull request merged**